ddocs/design-rafka-rethinking.rst

```diff
@@ -0,0 +1,210 @@
+`Producer` API, by taking advantage of the Redis protocol and leveraging the already existing Redis
+client libraries, mainly targeting languages that lack of a Kafka client library with proper
+concurrency primitives. `However, the initial design decision for a simple and straightforward API
+is not reflected in the internal structure of Rafka`.
+
+Shortly, in order to implement the (simple) Rafka API, we internally maintain `five` primary
+structs; the `Server`, `Client`, `Consumer`, `Producer`, and the `ConsumerManager`. Those structs
+work interchangeably by operating on common data structures, which apparently translates to a lot
+of extra lines of code for synchronizing and properly orchestrating the requests handling. Besides
+the error-prone nature of that extra complexity, it proved to be quite difficult to follow the
+execution flow and most important to isolate the interested parties in case of a bug or failure.
+
+Furthermore, an important feature that Rafka currently lacks of, is the support of proper
+monitoring primitives and metrics. There's no way to monitor the server's status, inspect the
+Rafka-to-Kafka interaction, follow a request's life-cycle, and generally monitoring the server's
+performance. There are a lot of incidents reported lately about lost messages on high-traffic
+production setups and it is currently impossible to identify the root cause of those failures.
+
+
+Background
+==========
+
+This section provides some extra information about Rafka internals, in order to facilitate the
+reading of the rest of this design document.
+
+Core components
+---------------
+
+The majority of Rafka functionality is internally handled by the following `5` major components. A
+short description of each of those components, follows:
+
+* ``Producer``, the interface used to send data streams to Kafka. It is allowed to produce to any
+  Kafka topic using the same `Producer` instance.
+
+* ``Consumer``, it allows reading data streams from Kafka. A `Consumer` is allowed to be subscribed
+  to either one topic or a list of Kafka topics.
+
+* ``ConsumerManager``, it maintains a mapping of **all** registered `Consumers` in the server. It
+  uses this mapping to return the appropriate `Consumer` to the requested `Client`. There's a
+  single `ConsumerManager` object per server and so it's scope is global.
+
+* ``Client``, it is created upon a new connection to Rafka and can currently serve a `Producer`
+  and/or one (or more) `Consumer` processes. It is associated with a client ID which apparently is
+  unique across the server (this is ensured via the ``ClientByID`` server's field). Also note that
+  each `Client` instance holds a reference to the global `ConsumerManager` object, too.
+
+* ``Server``, it maintains a map with **all** registered `Clients`. It holds a reference to the
+  global `ConsumerManager` object, just like the registered `Clients` do. It also takes over the
```

Comment:
"just like the registered `Clients` do", maybe this should change to "just like the registered `Consumers` do"?

Reply:
Nope, its `Clients`, not `Consumers`!